### ISSUE TITLE
Don't use IPv6 when NICs have only local IPv6 addresses

### ISSUE THREAD

**Issue by yetanotherion** [twitter/finagle#852]
When NICs have only local IPv6 address-es, finagle fails to reach destination hosts.
A fix for finagle was proposed at https://github.com/twitter/finagle/pull/849
Netty fixed the issue at https://github.com/netty/netty/pull/10170

### Expected behavior

Finagle does not use the IPv6 address-es returned by AAAA queries.

### Actual behavior
When NICs have IPv6 address-es configured, even if they're only local, finagle sends both AAAA and A DNS queries.

If the AAAA query gets an answer, it (sometimes, most probably if the AAAA is answered faster than A) decides to use the IPv6 address of the host the client is trying to contact.
Unfortunately, given that there are only local address-es, finagle won't be able to reach the host with IPv6.

### Steps to reproduce the behavior
Configure a host having local IPv6 address-es only
https://github.com/netty/netty/pull/10170

```scala
import com.twitter.finagle.service.RetryBudgets
import com.twitter.finagle.{Http, Service, http}
import com.twitter.util.{Await, Future}

object Main extends App {
  val host = "www.google.com:80"
  val clientStack = StackClient.newStack
  val client: Service[Request, Response] =
    Http
      .Client(StackClient.newStack)
      .withRetryBudget(RetryBudgets.EMPTY)
      .newService(host)
  val response: Future[http.Response] = client(
    http.Request(http.Method.Get, "")
  )
  val result = response.onSuccess { rep =>
    println(rep.toString())
  }
  Await.result(result)
}
```

```shell
$ java -jar debug-ipv6-uber.jar
May 18, 2020 7:59:41 AM com.twitter.finagle.Init$$anonfun$6 apply$mcV$sp
INFO: Finagle version 20.4.0 (rev=3775be321525bc2668dbbc379e90a937f7b8a680) built at 20200402-044802
Response("HTTP/1.1 Status(200)")
$ java -jar debug-ipv6-uber.jar
May 18, 2020 7:59:44 AM com.twitter.finagle.Init$$anonfun$6 apply$mcV$sp
INFO: Finagle version 20.4.0 (rev=3775be321525bc2668dbbc379e90a937f7b8a680) built at 20200402-044802
Exception in thread "main" Failure(null: www.google.com/2a00:1450:4007:808:0:0:0:2004:80 at remote address: www.google.com/2a00:1450:4007:808:0:0:0:2004:80. Remote Info: Not Available, flags=0x18) with RemoteInfo -> Upstream Address: Not Available, Upstream id: Not Available, Downstream Address: www.google.com/2a00:1450:4007:808:0:0:0:2004:80, Downstream label: www.google.com:80, Trace Id: 9b4eea10d26b65fa.9b4eea10d26b65fa<:9b4eea10d26b65fa with Service -> www.google.com:80
Caused by: com.twitter.finagle.ConnectionFailedException: null: www.google.com/2a00:1450:4007:808:0:0:0:2004:80 at remote address: www.google.com/2a00:1450:4007:808:0:0:0:2004:80. Remote Info: Not Available
	at com.twitter.finagle.netty4.ConnectionBuilder$$anon$1.operationComplete(ConnectionBuilder.scala:105)
	at com.twitter.finagle.netty4.ConnectionBuilder$$anon$1.operationComplete(ConnectionBuilder.scala:84)
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:577)
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:551)
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:490)
	at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:615)
	at io.netty.util.concurrent.DefaultPromise.setFailure0(DefaultPromise.java:608)
	at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:117)
	at io.netty.channel.epoll.AbstractEpollChannel$AbstractEpollUnsafe.connect(AbstractEpollChannel.java:599)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.connect(DefaultChannelPipeline.java:1342)
	at io.netty.channel.AbstractChannelHandlerContext.invokeConnect(AbstractChannelHandlerContext.java:548)
	at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:533)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.connect(CombinedChannelDuplexHandler.java:495)
	at io.netty.channel.ChannelOutboundHandlerAdapter.connect(ChannelOutboundHandlerAdapter.java:51)
	at io.netty.channel.CombinedChannelDuplexHandler.connect(CombinedChannelDuplexHandler.java:296)
	at io.netty.channel.AbstractChannelHandlerContext.invokeConnect(AbstractChannelHandlerContext.java:548)
	at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:533)
	at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:517)
	at io.netty.channel.DefaultChannelPipeline.connect(DefaultChannelPipeline.java:978)
	at io.netty.channel.AbstractChannel.connect(AbstractChannel.java:253)
	at io.netty.bootstrap.Bootstrap$3.run(Bootstrap.java:250)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:387)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at com.twitter.finagle.util.BlockingTimeTrackingThreadFactory$$anon$1.run(BlockingTimeTrackingThreadFactory.scala:23)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: io.netty.channel.AbstractChannel$AnnotatedNoRouteToHostException: null: www.google.com/2a00:1450:4007:808:0:0:0:2004:80
Caused by: java.net.NoRouteToHostException
...
```

**Comment by yetanotherion** [comment]
Hello @bryce-anderson , @hamdiallam , @mosesn , @roanta, 
this is a kind reminder of this issue posted the 18th May. https://github.com/twitter/finagle/issues/857 was created the 13th of June and
got a feedback from one of you. This could indicate that this issue may have been considered as having a lower priority, or that it did not fit the process depicted in https://github.com/twitter/finagle/blob/develop/CONTRIBUTING.md. After reading that file again, I don't yet see what was missed. Still, I'll be glad to refactor as needed. In any case, feedbacks will be very much appreciated.

For more context, in the process to making our infra IPv6 compatible, we're trying to remove all _**-Djava.net.preferIPv4Stack=true**_ that are currently mandatory for all programs using finagle in our systems. We do think users of finagle migrating to IPv6 will face the same issue.
To sum up, this issue and the related https://github.com/twitter/finagle/pull/849 aim at smoothing the migration to IPv6.

**Comment by hamdiallam** [comment]
I'll loop back to the PR this week. The changes looked fine on the initial pass. Sorry for missing this!

**Comment by hamdiallam** [comment]
looking now
